### PR TITLE
[FEAT][ROCm]: Support AITER MLA on V1 Engine

### PR DESCRIPTION
**Pull request by vllmellm**
## AITER MLA Support for V1 Engine

This PR implements AITER MLA attention backend support for the V1 engine. The implementation mirrors the V0 engine's established approach from [PR #15893](https://github.com/vllm-project/vllm/pull/15893).

This PR also introduces a new environment variable, `VLLM_ROCM_EXECUTE_MODEL_TIMEOUT`, which specifies the model execution timeout in seconds. This allows for flexible adjustment of execution time, which is helpful since a timeout error was encountered during graph building when enabling AITER MLA ops on the v1 engine.

### Accuracy Validation
using the command below:
`VLLM_ATTENTION_BACKEND=ROCM_AITER_MLA VLLM_USE_V1=1 lm_eval \
  --model vllm \
  --model_args pretrained=deepseek-ai/DeepSeek-V3,tensor_parallel_size=8,trust_remote_code=True,max_model_len=32768,block_size=1,enforce_eager=False \
  --tasks gsm8k  --num_fewshot 5 --batch_size auto `

Results:
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9492|±  |0.0060|
|     |       |strict-match    |     5|exact_match|↑  |0.9477|±  |0.0061|

### Performance:
The results of `benchmarks/benchmark_serving.py`

using the commands below:
v0 engine = `VLLM_ATTENTION_BACKEND=ROCM_AITER_MLA VLLM_USE_V1=0 python benchmarks/benchmark_serving.py --model deepseek-ai/DeepSeek-V3  --trust-remote-code --dataset-name random`
v1 engine = `VLLM_ATTENTION_BACKEND=ROCM_AITER_MLA VLLM_USE_V1=1 python benchmarks/benchmark_serving.py --model deepseek-ai/DeepSeek-V3  --trust-remote-code --dataset-name random`

| Metric                                       | ROCm AITER MLA V1                   | ROCm AITER MLA V0 |
|----------------------------------------------|-------------------------------------|-------------------|
| Successful requests                          | 1000                                | 1000              |
| Benchmark duration (s)                       | 179.78                              | 180.92            |
| Total input tokens                           | 1024000                             | 1024000           |
| Total generated tokens                       | 39667                               | 38739             |
| Request throughput (req/s)                   | 5.56                                | 5.53              |
| Output token throughput (tok/s)              | 220.64                              | 214.12            |
| Total Token throughput (tok/s)               | 5916.41                             | 5873.93           |
|----------------------------------------------|-------------------------------------|-------------------|
| Mean TTFT (ms)                               | 85618.95                            | 87984.79          |
| Median TTFT (ms)                             | 85528.56                            | 91664.85          |
| P99 TTFT (ms)                                | 166979.65                           | 96858.61          |
|----------------------------------------------|-------------------------------------|-------------------|
| Mean TPOT (ms)                               | 1048.37                             | 4727.25           |
| Median TPOT (ms)                             | 1242.70                             | 1189.05           |
| P99 TPOT (ms)                                | 1593.63                             | 38742.17          |
|----------------------------------------------|-------------------------------------|-------------------|
| Mean ITL (ms)                                | 698.66                              | 857.94            |
| Median ITL (ms)                              | 1226.67                             | 110.7             |
| P99 ITL (ms)                                 | 1664.92                             | 10027.72          |

### CODE REVIEW

File: vllm/engine/arg_utils.py [hunk 1]
```diff
@@ -1298,6 +1298,8 @@ def _is_v1_supported_oracle(self, model_config: ModelConfig) -> bool:
             "FLASHMLA",
             "FLASHINFER",
             "FLASHINFER_VLLM_V1",
+            "ROCM_AITER_MLA",
+            "ROCM_AITER_MLA_VLLM_V1",
```
Comment (houseroad) [inline vllm/engine/arg_utils.py:1302]: it seems not necessary to have ROCM_AITER_MLA_VLLM_V1 backend. But it seems benign here.

**Review by houseroad — APPROVED** [review]
Looks good to me.

File: vllm/attention/ops/rocm_aiter_mla.py [hunk 2]
```diff
@@ -40,3 +65,24 @@ def aiter_mla_decode_fwd(
...
+    logit_cap: float = 0.0,
+) -> None:
+    pass
+
+
+if current_platform.is_rocm():
+    direct_register_custom_op(op_name="rocm_aiter_mla_decode_fwd",
+                              op_func=mla_decode_fwd_impl,
+                              mutates_args=["o"],
+                              fake_impl=mla_decode_fwd_fake,
```
[the remark this answers is not in the document]
Comment (vllmellm) [reply vllm/attention/ops/rocm_aiter_mla.py:87]: @houseroad Yes without it there is error from torch dynamo while building cuda graphs.

File: vllm/engine/arg_utils.py (hunk 1 above)
Comment (vllmellm) [reply vllm/engine/arg_utils.py:1302]: @houseroad thanks for pointing out. removed `ROCM_AITER_MLA_VLLM_V1` this from list.

File: vllm/v1/executor/multiproc_executor.py [hunk 3]
```diff
@@ -38,7 +40,8 @@
 POLLING_TIMEOUT_MS = 5000
 POLLING_TIMEOUT_S = POLLING_TIMEOUT_MS // 1000
 
-EXECUTE_MODEL_TIMEOUT_S = 40
+EXECUTE_MODEL_TIMEOUT_S = (envs.VLLM_ROCM_EXECUTE_MODEL_TIMEOUT
```
Comment (houseroad) [inline vllm/v1/executor/multiproc_executor.py:43]: wondering why rocm needs a much larger timeout here?

**Comment by houseroad** [comment]
Could you check the failed tests?

**Review by hongxiayang — APPROVED** [review]
Approve with comment.

To make Deepseek V1 performant, it needs additional work.

Based on my test, it can improve TTFT if additional AITER environment variables are used. Otherwise, the TTFT is not as good comparing to V0. Throughput is not good yet comparing to V0.

File: vllm/platforms/rocm.py [hunk 4]
```diff
@@ -155,10 +155,22 @@ def get_attn_backend_cls(cls, selected_backend, head_size, dtype,
...
-                    return "vllm.attention.backends.rocm_aiter_mla.AiterMLABackend"  # noqa: E501
+                    if use_v1:
+                        logger.info("Using AITER MLA backend on V1 engine.")
+                        logger.warning(
+                            "Increasing the model execution timeout"
+                            "using the VLLM_ROCM_EXECUTE_MODEL_TIMEOUT"
+                            "environment variable is recommended"
+                            "if timeout error is encountered"
+                            "when running %s"
+                            "backend on V1 engine.", selected_backend)
```
Comment (hongxiayang) [inline vllm/platforms/rocm.py:169]:
```suggestion
                            "Increasing the model execution timeout "
                            "using the VLLM_ROCM_EXECUTE_MODEL_TIMEOUT "
                            "environment variable is recommended "
                            "if timeout error is encountered "
                            "when running %s "
                            "backend on V1 engine.", selected_backend)
```

File: vllm/v1/executor/multiproc_executor.py (hunk 3 above)
Comment (vllmellm) [reply vllm/v1/executor/multiproc_executor.py:43]: on first time run when the graph is being created it might take between 100-250 seconds depending on how many AITER kernels are enabled. Thus we kept the default timeout to 250s.

File: vllm/attention/backends/rocm_aiter_mla.py [hunk 5]
```diff
@@ -16,7 +16,7 @@
 from vllm.attention.backends.utils import (compute_slot_mapping,
                                            compute_slot_mapping_start_idx,
                                            is_block_tables_empty)
-from vllm.attention.ops.rocm_aiter_mla import (aiter_mla_decode_fwd,
+from vllm.attention.ops.rocm_aiter_mla import (aiter_mla_decode_forward,
```
[the remark this answers is not in the document]
Comment (vllmellm) [reply vllm/attention/backends/rocm_aiter_mla.py]: it has been addressed in the latest commit.

File: vllm/v1/attention/backends/mla/rocm_aiter_mla.py [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+        if any(unsupported_features):
+            raise NotImplementedError(
+                "Aiter MLA does not support one of the following: "
+                "alibi_slopes, sliding_window, blocksparse_params, "
+                "logits_soft_cap")
+
+        from aiter import flash_attn_varlen_func
+        self.flash_attn_varlen_func = flash_attn_varlen_func
+
+    def _flash_attn_varlen_diff_headdims(self,
```
Comment (SageMoore) [inline vllm/v1/attention/backends/mla/rocm_aiter_mla.py:152]: Where is this used?

File: vllm/v1/attention/backends/mla/rocm_aiter_mla.py [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+
+class AiterMLAMetadataBuilder(MLACommonMetadataBuilder[AiterMLAMetadata]):
+
+    def __init__(self, runner):
+        super().__init__(runner)
+        max_model_len = self.runner.model_config.max_model_len
+        assert max_model_len == 32768,\
+            "AITER MLA requires max_model_len=32768"
+        assert self.runner.block_size == 1, "AITER MLA" \
+            "requires only block size 1."
```
Comment (SageMoore) [inline vllm/v1/attention/backends/mla/rocm_aiter_mla.py:68]: Nit: "only supports block size 1."

File: vllm/v1/executor/multiproc_executor.py (hunk 3 above)
Comment (SageMoore) [reply vllm/v1/executor/multiproc_executor.py:43]: I'm not crazy about requiring another environment variable when running AITER. Can you just set the timeout to 250 here instead of asking the user to increase the timeout? Feel free to give a "safe" timeout.

File: vllm/envs.py [hunk 8]
```diff
@@ -488,6 +489,10 @@ def maybe_convert_int(value: Optional[str]) -> Optional[int]:
     "VLLM_RPC_TIMEOUT":
     lambda: int(os.getenv("VLLM_RPC_TIMEOUT", "10000")),
 
+    # Time in seconds for the model execution in ROCm platforms.
+    "VLLM_ROCM_EXECUTE_MODEL_TIMEOUT":
```
Comment (SageMoore) [inline vllm/envs.py:493]: Let's remove this. See below comment.

File: vllm/v1/attention/backends/mla/common.py [hunk 9]
```diff
@@ -496,11 +496,12 @@ def build(self, num_reqs: int, num_actual_tokens: int, max_query_len: int,
                 max_context_chunk = (self.chunked_prefill_workspace_size //
                                      num_prefills_with_context_cpu)
 
-                # align max_context_chunk to page_size by rounding down,
-                # currently the `gather_cache` kernel cannot handle
-                # `context_chunk_starts` that are not aligned to page_size
-                max_context_chunk = round_down(max_context_chunk,
-                                               self.page_size)
+                if self.aot_schedule:
```
Comment (SageMoore) [inline vllm/v1/attention/backends/mla/common.py:499]: Could you explain this a bit? Why was this change necessary?

**Review by SageMoore — CHANGES_REQUESTED** [review]
Looks reasonable. Just a few nits and questions

File: vllm/v1/attention/backends/mla/common.py (hunk 9 above)
Comment (vllmellm) [reply vllm/v1/attention/backends/mla/common.py:499]: @SageMoore the `self.page_size` if only defined in `__init__` with the condition `self.aot_schedule` while on ROCm this condition is not true and it encounters the error self.page_size is not defined. if it meant not to be used on AMD then should not be used as there is a comment that points out here [in this line] (https://github.com/vllm-project/vllm/blob/ba7703e659e21ea376c25b872c4a80a7c82804fe/vllm/v1/attention/backends/mla/common.py#L358) to not access runner to on AMD thus have to ask the author. 
anyways if `self.page_size` is defined without this condition `self.aot_schedule ` it does not have any effect on ROCm at least for AITER MLA which is the only MLA backend in V1 currently.

File: vllm/v1/attention/backends/mla/rocm_aiter_mla.py (hunk 6 above)
Comment (vllmellm) [reply vllm/v1/attention/backends/mla/rocm_aiter_mla.py:152]: @SageMoore you may want to check `coomon.py`  the method `_flash_attn_varlen_diff_headdims` is defined there and overridden in this class.

File: vllm/envs.py (hunk 8 above)
Comment (vllmellm) [reply vllm/envs.py:493]: @SageMoore At this moment we can't find "safe" timeout because depending on number of AITER kernels are enable knowing the "safe" timeout is difficult as tracing the AITER jit files might be time consuming during execution time and might change as AITER ops might change based on different versions being used in future. Thus, rather than having a hardcoded timeout that might trouble the user where to spot it in the code they are able to control this value with environment variable.

File: vllm/v1/attention/backends/mla/rocm_aiter_mla.py (hunk 6 above)
Comment (SageMoore) [reply vllm/v1/attention/backends/mla/rocm_aiter_mla.py:152]: I see now. I must have mistyped the string when I searched for it :).

File: vllm/envs.py (hunk 8 above)
Comment (SageMoore) [reply vllm/envs.py:493]: I see. No problem.

File: vllm/v1/executor/multiproc_executor.py (hunk 3 above)
Comment (tlrmchlsmth) [reply vllm/v1/executor/multiproc_executor.py:43]: Will these timeouts only happen during the profile and graph capture runs? Or can they happen while processing real requests?

File: vllm/envs.py (hunk 8 above)
Comment (vllmellm) [reply vllm/envs.py]: @SageMoore The environment variable has been removed. The current default value is set to "250".

Comment (hongxiayang) [reply vllm/envs.py]: @SageMoore : the env change is removed as we discussed. Please merge this asap if there are no other blockers.

**Review by SageMoore — APPROVED** [review]
Looks reasonable. Thanks for taking out the timeout changes!

**Comment by chaunceyjiang** [comment]
It seems that this PR introduced a static check error.

https://github.com/vllm-project/vllm/actions/runs/14923384538/job/41922811885?pr=17845
```
Error: vllm/v1/attention/backends/mla/rocm_aiter_mla.py:98: error: Signature of "_build_decode" incompatible with supertype "MLACommonMetadataBuilder"  [override]
vllm/v1/attention/backends/mla/rocm_aiter_mla.py:98: note:      Superclass:
vllm/v1/attention/backends/mla/rocm_aiter_mla.py:98: note:          def _build_decode(self, block_table: Any, seq_lens: Any) -> Any
vllm/v1/attention/backends/mla/rocm_aiter_mla.py:98: note:      Subclass:
vllm/v1/attention/backends/mla/rocm_aiter_mla.py:98: note:          def _build_decode(self, input_positions: Any, block_table: Any, seq_lens: Any) -> AiterMLADecodeMetadata
Error: vllm/v1/attention/backends/mla/rocm_aiter_mla.py:108: error: Unexpected keyword argument "input_positions" for "AiterMLADecodeMetadata"  [call-arg]
Found 2 errors in 1 file (checked 83 source files)

```

**Comment by tjtanaa** [comment]
> EXECUTE_MODEL_TIMEOUT_S

fixed by https://github.com/vllm-project/vllm/pull/17880
